1-Parallelize-your-python-code_Futures_API.ipynb

```diff
@@ -0,0 +1,893 @@
+   "source": [
+    "future"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "72bf4e6a-8553-4383-b15a-c9f249bd71df",
+   "metadata": {},
+   "source": [
+    "You can block on the computation and gather the result with the `.result()` method."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7f9bf670-2d56-4a7a-bdb4-36791b91de05",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "future.result()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "202e4b5d-b1c3-49f2-b5e7-5cc02602a4e4",
+   "metadata": {},
+   "source": [
+    "**`client.map()`**"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "4c5576bf-3fdd-4fed-9419-a28ca323c850",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "futures = client.map(inc, range(8))  # returns immediately with pending list of futures\n",
+    "futures"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f94b34a6-fbd4-4ede-a9c0-9b90e71f4e1b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "future_sum = client.submit(sum, futures)\n",
+    "future_sum.result()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "8248e9e5-08b5-496d-9c51-94119afed9fe",
+   "metadata": {},
+   "source": [
+    "**`as_completed()`**"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "35499d7d-ec10-44db-8fbd-a7a061b3e012",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from dask.distributed import as_completed\n",
+    "import numpy as np"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "4eb4ac56-bdee-4c6f-a8a8-6b66c12a974e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def random_score(x):\n",
+    "    return np.random.uniform(low=0.5, high=10.0)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "dbb2fc13-51be-43a0-b47a-b89db510ff6a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "score_futures = client.map(random_score, range(20))\n",
+    "\n",
+    "best_score = 0\n",
+    "for future in as_completed(score_futures):\n",
+    "    score = future.result()\n",
+    "    if score > best_score:\n",
+    "        best_score = score"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "606792cf-422f-45bc-a570-bea637dfb4f7",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "print(best_score)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d7be9fac-d3a0-4150-9d45-4e2bb787abf8",
+   "metadata": {},
+   "source": [
+    "**Useful links: futures**\n",
+    "* [Futures documentation](https://docs.dask.org/en/latest/futures.html)\n",
+    "* [Futures screencast](https://www.youtube.com/watch?v=07EiCpdhtDE)\n",
+    "* [Futures examples](https://examples.dask.org/futures.html)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "16467916-008c-4a21-bb1e-d374c530ee57",
+   "metadata": {},
+   "source": [
```

Comment:
In a recap (or maybe at the beginnning) I recommend adding content about why they might use futures.  What are common use cases?  Why might they use it over delayed?  etc..

I think that the message here is something like 

-  Futures are commonly used in simple situations

    Often you just want a parallel for loop and futures give that to you.   It might be a good idea to give them such an exercise before having them launch into something complex.  They're likely to get a bunch of things wrong and it would be good to clear those up before you get into anything complicated.
    
    One suggestion would be to continue to use the scraping example, but give them a list of webpages to scrape and a function that scrapes them, and just have them call that function many times in a simple for loop.  

- Futures are very flexible and can do lots of sophisticated things

    This is where I would introduce nested for loops, as_completed with crawling, and so on

Right now this jumps from toy examples into an exercise that I think will prove challenging to folks.  It might work, but I would anticiapte some pain.

Reply:
I modify all the simple examples to be in the context of the main example. that way we increase the complexity as we go, keeping the example consistent, and makes the transition smoother